### PR TITLE
Add a batch mpsc and upgrade crossbeam-channel to 0.3

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

## What have you changed? (mandatory)
Add a mpsc channel to support batch. It means 2 things:
1. The sender won't always notify the receiver for every message.
2. We have a new `BatchReceiver` to do some custom batch for received messages.

## What are the type of the changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
Improvement.

## How has this PR been tested? (mandatory)
make dev.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No.

## Does this PR affect tidb-ansible update? (mandatory)
No.

### CODE REVIEW

File: src/util/mpsc_batch.rs [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+use futures::{Async, Poll, Stream};
+
+const DEFAULT_NOTIFY_BATCH_SIZE: usize = 16;
+
+struct State {
+    sender_cnt: AtomicIsize,
+    receiver_cnt: AtomicIsize,
+    recv_task: AtomicOption<Task>,
+    notify_batch_size: usize,
+    un_notified: AtomicUsize,
```
Comment (overvenus) [inline src/util/mpsc_batch.rs]: How about `pending`？

**Comment by breeswish** [comment]
How about using https://docs.rs/crossbeam-channel/0.1.2/crossbeam_channel/struct.TryIter.html?

**Comment by MyonKeminta** [comment]
Emmmm, could you please give some comments or rustdoc to explain the responsibility of each parts?

File: src/util/mod.rs [hunk 2]
```diff
@@ -40,6 +40,7 @@ pub mod jemalloc;
 pub mod logger;
 pub mod metrics;
 pub mod mpsc;
+pub mod mpsc_batch;
```
Comment (BusyJay) [inline src/util/mod.rs]: Better put batch as submodule of mpsc.

File: src/util/mpsc_batch.rs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    #[inline]
+    fn inc_send_and_maybe_notify(&self) {
+        let pending = self.pending.fetch_add(1, Ordering::SeqCst);
+        if pending >= self.notify_batch_size - 1 {
+            self.do_notify();
+        }
+    }
+
+    #[inline]
+    fn do_notify(&self) {
```
Comment (zhangjinpeng1987) [inline src/util/mpsc_batch.rs]: s/do_notify/notify

File: src/util/mpsc/batch.rs [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+pub struct Receiver<T> {
+    receiver: crossbeam_channel::Receiver<T>,
+    state: Arc<State>,
+}
+
+impl<T> Sender<T> {
+    #[inline]
+    pub fn send(&self, t: T) -> Result<(), SendError<T>> {
+        if !self.state.is_receiver_closed() {
+            self.sender.send(t);
```
Comment (zhangjinpeng1987) [inline src/util/mpsc/batch.rs]: How about the receiver closed before this line but after the if?

**Comment by zhangjinpeng1987** [comment]
Please add a benchmark for this batched mpsc.

File: src/util/mpsc/batch.rs (hunk 4 above)
Comment (hicqu) [reply src/util/mpsc/batch.rs]: I will use crossbeam-channel 0.3 instead of 0.2 to resolve this.

**Comment by breeswish** [comment]
Maybe we should merge https://github.com/tikv/tikv/pull/3781 first to be more friendly for community contributors?

**Comment by hawkingrei** [comment]
If feasible, crossbeam can be upgraded to 5.0. it includes the crossbeam-channel、epoch and so on. so we can put crossbeam and crossbeam-channel into a lib.

**Comment by hicqu** [comment]
I have add a bench for batch receiver and unary receiver, here is the result:
```
test channel::bench_channel::bench_receiver_stream                    ... bench:   2,250,143 ns/iter (+/- 3,172,170)
test channel::bench_channel::bench_receiver_stream_batch              ... bench:   1,049,640 ns/iter (+/- 1,919,578)
```

File: src/util/mpsc/batch.rs [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+            recv_task: AtomicOption::new(),
+            notify_size,
+            old_pending: AtomicUsize::new(0),
+            notifier_registered: AtomicBool::new(false),
+        }
+    }
+
+    #[inline]
+    fn inc_send_and_maybe_notify(&self) {
+        let old_pending = self.old_pending.fetch_add(1, Ordering::SeqCst);
```
Comment (siddontang) [inline src/util/mpsc/batch.rs]: you can use AcqRel here

File: src/util/mpsc/batch.rs [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+        self.receiver.try_recv()
+    }
+
+    #[inline]
+    pub fn recv_timeout(&self, timeout: Duration) -> Result<T, RecvTimeoutError> {
+        self.receiver.recv_timeout(timeout)
+    }
+
+    #[inline]
+    fn yield_poll(&self) -> bool {
```
Comment (siddontang) [inline src/util/mpsc/batch.rs]: what does yield mean?

File: src/util/mpsc/batch.rs [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        State {
+            recv_task: AtomicOption::new(),
+            notify_size,
+            old_pending: AtomicUsize::new(0),
+            notifier_registered: AtomicBool::new(false),
+        }
+    }
+
+    #[inline]
+    fn inc_send_and_maybe_notify(&self) {
```
Comment (siddontang) [inline src/util/mpsc/batch.rs]: I prefer `try_notify` here

File: src/util/mpsc/batch.rs (hunk 6 above)
Comment (hicqu) [reply src/util/mpsc/batch.rs]: Just like yield instruction of CPU, or yield keywork in Python. Here it should means the `Receiver` will give up CPU and let Executors execute other futures.

File: src/util/mpsc/batch.rs [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+struct State {
+    recv_task: AtomicOption<Task>,
+    notify_size: usize,
+    // How many messages are sent without notify.
+    old_pending: AtomicUsize,
+    notifier_registered: AtomicBool,
+}
+
+impl State {
+    fn new(notify_size: usize) -> State {
```
Comment (zhangjinpeng1987) [inline src/util/mpsc/batch.rs]: assert!(notify_size > 0);

File: src/util/mpsc/batch.rs [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+            state: state.clone(),
+        },
+        Receiver { receiver, state },
+    )
+}
+
+/// Create a bounded channel with a given `notify_size`, which means if there are more pending
+/// messages in the channel than `notify_size`, the `Sender` will auto notify the `Receiver`.
+///
+/// ### Panics
```
Comment (overvenus) [inline src/util/mpsc/batch.rs]: Panic section should starts with a single `#`.

File: benches/misc/channel/bench_channel.rs [hunk 10]
```diff
@@ -169,7 +170,80 @@ fn bench_crossbeam_channel(b: &mut Bencher) {
...
 }
+
+#[bench]
+fn bench_receiver_stream_batch(b: &mut Bencher) {
+    let (tx, rx) = mpsc::batch::bounded::<i32>(128, 8);
+    for _ in 0..1 {
+        let tx1 = tx.clone();
+        thread::spawn(move || {
+            (0..usize::MAX)
+                .take_while(|i| tx1.send(*i as i32).is_ok())
```
[the remark this answers is not in the document]
Comment (hicqu) [reply benches/misc/channel/bench_channel.rs]: It's a bounded channel, so I guess there is no such issue here.

File: src/util/mpsc/batch.rs (hunk 8 above)
Comment (hicqu) [reply src/util/mpsc/batch.rs]: Already checked in `bounded` and `unbounded`.

**Comment by hicqu** [comment]
ping @zhangjinpeng1987 @MyonKeminta  PTAL when you are free, thanks!

**Pull request merged**